src/backend/distributed/planner/shard_pruning.c

```diff
@@ -0,0 +1,1267 @@
+	for (shardIndex = 0; shardIndex < length; shardIndex++)
+	{
+		ShardInterval *shardInterval =
+			shardArray[shardIndex];
+		shardIntervalList = lappend(shardIntervalList, shardInterval);
+	}
+
+	return shardIntervalList;
+}
+
+
+/*
+ * DeepCopyShardIntervalList copies originalShardIntervalList and the
+ * contained ShardIntervals, into a new list.
+ */
+static List *
+DeepCopyShardIntervalList(List *originalShardIntervalList)
+{
+	List *copiedShardIntervalList = NIL;
+	ListCell *shardIntervalCell = NULL;
+
+	foreach(shardIntervalCell, originalShardIntervalList)
+	{
+		ShardInterval *originalShardInterval =
+			(ShardInterval *) lfirst(shardIntervalCell);
+		ShardInterval *copiedShardInterval =
+			(ShardInterval *) palloc0(sizeof(ShardInterval));
+
+		CopyShardInterval(originalShardInterval, copiedShardInterval);
+		copiedShardIntervalList = lappend(copiedShardIntervalList, copiedShardInterval);
+	}
+
+	return copiedShardIntervalList;
+}
+
+
+/*
+ * PruneOne returns all shards in the table that match a single
+ * PruningInstance.
+ */
+static List *
+PruneOne(DistTableCacheEntry *cacheEntry, ClauseWalkerContext *context,
+		 PruningInstance *prune)
+{
+	ShardInterval *shardInterval = NULL;
+
+	/* Well, if life always were this easy... */
+	if (prune->evaluatesToFalse)
+	{
+		return NIL;
+	}
+
+	/*
+	 * For an equal constraints, if there's no overlapping shards (always the
+	 * case for hash and range partitioning, sometimes for append), can
+	 * perform binary search for the right interval.  That's usually the
+	 * fastest, so try that first.
+	 */
+	if (prune->equalConsts &&
+		!cacheEntry->hasOverlappingShardInterval)
+	{
+		shardInterval = FindShardInterval(prune->equalConsts->constvalue, cacheEntry);
+
+		/*
+		 * If pruned down to nothing, we're done. Otherwise see if other
+		 * methods prune down further / to nothing.
+		 */
+		if (!shardInterval)
+		{
+			return NIL;
+		}
+	}
+
+	/*
+	 * If the hash value we're looking for is known, we can search for the
+	 * interval directly.  That's fast and should only ever be the case for a
+	 * hash-partitioned table.
+	 */
+	if (prune->hashedEqualConsts)
+	{
+		int shardIndex = INVALID_SHARD_INDEX;
+		ShardInterval **sortedShardIntervalArray = cacheEntry->sortedShardIntervalArray;
+
+		Assert(context->partitionMethod == DISTRIBUTE_BY_HASH);
+
+		shardIndex = FindShardIntervalIndex(prune->hashedEqualConsts->constvalue,
+											cacheEntry);
+
+		if (shardIndex == INVALID_SHARD_INDEX)
+		{
+			return NIL;
+		}
+		else if (shardInterval &&
+				 sortedShardIntervalArray[shardIndex]->shardId != shardInterval->shardId)
+		{
+			return NIL;
+		}
+		else
+		{
+			return list_make1(sortedShardIntervalArray[shardIndex]);
+		}
+	}
+
+	/*
+	 * FIXME: We could perform ExhaustivePrune() on the individual
+	 * shardInterval, in case there's both an = and a < / >? It'd probably be
+	 * better to pre-compute that by checking a PruningInstance's
+	 * less[Equal]Consts/greater[Equal]Consts against equalConsts.
```

Comment:
This comment doesn't help me understand why we return below. Maybe we can say "We found a shard based on an equality constraint. We might be able to prune further if there are also range constraints, but this is atypical and we can save on planning cost by returning earlier."